File: core/src/main/java/com/worldreader/core/datasource/network/datasource/oauth/OAuthNetworkDataSource.java
package com.worldreader.core.datasource.network.datasource.oauth;

import com.worldreader.core.datasource.network.model.OAuthNetworkResponseEntity;
import com.worldreader.core.error.user.LoginException;

public interface OAuthNetworkDataSource {

  OAuthNetworkResponseEntity applicationToken();

  OAuthNetworkResponseEntity refreshToken(String refreshToken);

  OAuthNetworkResponseEntity userToken(String username, String password) throws LoginException;

  OAuthNetworkResponseEntity userTokenWithFacebook(String facebookToken) throws LoginException;

  OAuthNetworkResponseEntity userTokenWithGoogle(String googleId, String email)
      throws LoginException;
}

File: harmony-kotlin/src/commonMain/kotlin/com/harmony/kotlin/library/oauth/data/datasource/network/OAuthNetworkDataSource.kt
package com.harmony.kotlin.library.oauth.data.datasource.network

import com.harmony.kotlin.data.datasource.PutDataSource
import com.harmony.kotlin.data.query.Query
import com.harmony.kotlin.error.notSupportedQuery
import com.harmony.kotlin.library.oauth.data.datasource.network.model.OAuthBodyRequest
import com.harmony.kotlin.library.oauth.data.entity.OAuthTokenEntity
import com.harmony.kotlin.library.oauth.data.query.OAuthQuery
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType

internal class OAuthNetworkDataSource(
  private val httpClient: HttpClient,
  private val apiPath: String,
  private val basicAuthorizationCode: String
) : PutDataSource<OAuthTokenEntity> {

  override suspend fun put(query: Query, value: OAuthTokenEntity?): OAuthTokenEntity {
    val bodyRequest = when (query) {
      is OAuthQuery.Password -> OAuthBodyRequest.Password(query.username, query.password)
      is OAuthQuery.RefreshToken -> OAuthBodyRequest.RefreshToken(query.refreshToken)
      is OAuthQuery.ClientCredentials -> OAuthBodyRequest.ClientCredentials(
        query.clientId,
        query.clientSecret
      )

      else -> notSupportedQuery()
    }

    val url = "$apiPath/auth/token"
    return httpClient.post(url) {
      header("Authorization", "Basic $basicAuthorizationCode")
      contentType(ContentType.Application.Json)
      setBody(bodyRequest)
    }.body()
  }
}
